### PR TITLE
X-HTTP-Method-Override listener

### PR DESCRIPTION
**Pull request by Wilt**
X-HTTP-Method-Override listener feature.
For this issue: https://github.com/zfcampus/zf-apigility/issues/175#issuecomment-241523026

### CODE REVIEW

File: test/HttpMethodOverrideListenerTest.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        $request->setMethod('GET');
+        $request->getHeaders()->addHeaderLine('X-HTTP-Method-Override', $method);
+
+        $event = new MvcEvent();
+        $event->setRequest($request);
+        $event->setRouteMatch($this->createRouteMatch([]));
+
+        $result = $listener($event);
+        $this->assertEquals($method, $request->getMethod());
+
```
Comment (webimpress) [inline test/HttpMethodOverrideListenerTest.php]: Please remove empty line above.

File: src/HttpMethodOverrideListener.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+        if(!$request->getHeaders()->has('X-HTTP-Method-Override')){
+            return;
+        }
+
+        $header = $request->getHeader('X-HTTP-Method-Override');
+
+        $method = $header->getFieldValue();
+
+        if (!in_array($method, $this->methods)) {
```
Comment (webimpress) [inline src/HttpMethodOverrideListener.php]: Please add one space after `!` op.

File: src/HttpMethodOverrideListener.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (webimpress) [inline src/HttpMethodOverrideListener.php]: Please update year - `2016` instead of `2014` here.

File: test/HttpMethodOverrideListenerTest.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (webimpress) [inline test/HttpMethodOverrideListenerTest.php]: Please update year - `2016` instead of `2014`.

File: test/HttpMethodOverrideListenerTest.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+        $request = new HttpRequest();
+        $request->setMethod('GET');
+        $request->getHeaders()->addHeaderLine('X-HTTP-Method-Override', 'TEST');
+
+        $event = new MvcEvent();
+        $event->setRequest($request);
+
+        $result = $listener($event);
+        $this->assertInstanceOf('ZF\ApiProblem\ApiProblemResponse', $result);
```
Comment (webimpress) [inline test/HttpMethodOverrideListenerTest.php]: Please use `::class` notation

File: src/HttpMethodOverrideListener.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        $header = $request->getHeader('X-HTTP-Method-Override');
+
+        $method = $header->getFieldValue();
+
+        if (!in_array($method, $this->methods)) {
+            return new ApiProblemResponse(new ApiProblem(
+                400,
+                'unrecognized method in X-HTTP-Method-Ovverride header'
```
Comment (webimpress) [inline src/HttpMethodOverrideListener.php]: Please change to uppercase `U` in `unrecognized` word

**Comment by Wilt** [comment]
PR for this issue: https://github.com/zfcampus/zf-apigility/issues/175#issuecomment-241523026

Didn't yet implement a solution for connecting the listener. Maybe an optional config param for enabling this listener. I think enabling this listener should be optional, feels not good to connect such a listener by default to the application.

Maybe someone from zf-campus can give some guidance on how they would like to connect it...?

File: src/HttpMethodOverrideListener.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        $header = $request->getHeader('X-HTTP-Method-Override');
+
+        $method = $header->getFieldValue();
+
+        if (! in_array($method, $this->methods)) {
+            return new ApiProblemResponse(new ApiProblem(
+                400,
+                'unrecognized method in X-HTTP-Method-Ovverride header'
```
Comment (webimpress) [inline src/HttpMethodOverrideListener.php]: `U` in `unrecognized` please

File: test/HttpMethodOverrideListenerTest.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        $event->setRequest($request);
+        $event->setRouteMatch($this->createRouteMatch([]));
+
+        $result = $listener($event);
+        $this->assertEquals($method, $request->getMethod());
+    }
+
+    /**
+     *
+     */
```
Comment (webimpress) [inline test/HttpMethodOverrideListenerTest.php]: What does it for?

File: test/HttpMethodOverrideListenerTest.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        $request->getHeaders()->addHeaderLine('X-HTTP-Method-Override', 'TEST');
+
+        $event = new MvcEvent();
+        $event->setRequest($request);
+
+        $result = $listener($event);
+        $this->assertInstanceOf(ApiProblemResponse::class, $result);
+        $problem = $result->getApiProblem();
+        $this->assertEquals(400, $problem->status);
+        $this->assertContains('unrecognized method in X-HTTP-Method-Ovverride header', $problem->detail);
```
Comment (webimpress) [inline test/HttpMethodOverrideListenerTest.php]: `U` in `unrecognized` please

**Comment by Wilt** [comment]
Updated the PR:

- Made HttpMethodOverrideListener extend AbstractListenerAggregate
- Attached listener depending on `x_http_method_override_enabled`
- Added $xHttpMethodOverrideEnabled to ContentNegotiationOptions (defaults to false);

**Comment by snapshotpl** [comment]
@Wilt there is already existing module https://github.com/rstgroup/http-method-override

**Comment by Wilt** [comment]
@snapshotpl OK, thanks. 
If there is no interest to implement this in Apigility/zf-content-negotiation, then someone can just close and I will delete my PR.

File: src/Factory/HttpMethodOverrideListenerFactory.php [hunk 10]
```diff
@@ -0,0 +1,27 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014-2016 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (webimpress) [inline src/Factory/HttpMethodOverrideListenerFactory.php]: Please use only `2016` year in new files.

File: test/Factory/HttpMethodOverrideListenerFactoryTest.php [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        $serviceManager = new ServiceManager();
+        $serviceManager->setService(
+            ContentNegotiationOptions::class,
+            new ContentNegotiationOptions()
+        );
+
+        $factory = new HttpMethodOverrideListenerFactory();
+
+        $service = $factory($serviceManager, 'HttpMethodOverrideListener');
```
Comment (webimpress) [inline test/Factory/HttpMethodOverrideListenerFactoryTest.php]: Please use here as second parameter `HttpMethodOverrideListener::class` (I know, it is not used)

File: test/Factory/HttpMethodOverrideListenerFactoryTest.php [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+class HttpMethodOverrideListenerFactoryTest extends TestCase
+{
+    public function testCreateServiceShouldReturnContentTypeFilterListenerInstance()
+    {
+        $serviceManager = new ServiceManager();
+        $serviceManager->setService(
+            ContentNegotiationOptions::class,
+            new ContentNegotiationOptions()
+        );
```
Comment (webimpress) [inline test/Factory/HttpMethodOverrideListenerFactoryTest.php]: You can use `->prophesize(...)` to create service manager mock (please check how it is done in other tests)

File: test/HttpMethodOverrideListenerTest.php [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    protected $httpMethodOverride = [
+        HttpRequest::METHOD_GET => [
+            HttpRequest::METHOD_HEAD,
+            HttpRequest::METHOD_POST,
+            HttpRequest::METHOD_PUT,
+            HttpRequest::METHOD_DELETE,
+            HttpRequest::METHOD_PATCH,
+        ],
+        HttpRequest::METHOD_POST => [
+        ]
```
Comment (webimpress) [inline test/HttpMethodOverrideListenerTest.php]: Trailing comma please.

File: src/HttpMethodOverrideListener.php [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    public function __construct(array $httpMethodOverride)
+    {
+        $this->httpMethodOverride = $httpMethodOverride;
+    }
+
+    /**
+     * Priority is set very high (should be executed before all other listeners that rely on the request method value).
+     * TODO: Check priority value, maybe value should be even higher??
+     *
+     * @param  EventManagerInterface $events
```
Comment (webimpress) [inline src/HttpMethodOverrideListener.php]: one space after `@param` please

File: test/Factory/HttpMethodOverrideListenerFactoryTest.php [hunk 15]
```diff
@@ -0,0 +1,31 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (webimpress) [inline test/Factory/HttpMethodOverrideListenerFactoryTest.php]: `2016` year please

**Comment by webimpress** [comment]
@Wilt I like it, I think will be nice to have this functionality out-of-the-box. 👍

**Comment by Wilt** [comment]
@webimpress Thanks for the feedback. 

I changed to prophecy. I don't like the fact that the `get` method and `getHttpOverrideMethods` in my test aren't recognized. Is there a workaround?

File: test/HttpMethodOverrideListenerTest.php [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    protected $httpMethodOverride = [
+        HttpRequest::METHOD_GET => [
+            HttpRequest::METHOD_HEAD,
+            HttpRequest::METHOD_POST,
+            HttpRequest::METHOD_PUT,
+            HttpRequest::METHOD_DELETE,
+            HttpRequest::METHOD_PATCH
+        ],
+        HttpRequest::METHOD_POST => [
+        ]
```
Comment (webimpress) [inline test/HttpMethodOverrideListenerTest.php]: please add trailing comma

**Comment by webimpress** [comment]
@Wilt do you mean your IDE doesn't recognize these methods?
You can always add comment `/** @var YourClass $variable */` before variable, and if you are using PHPStorm it will be recognized properly.
But then it's not 100% true, because prophecy object is `ProphecyInterface`, so maybe better if your comment will be:
`/** @var YourClass|ProphecyInterface $variable */`

**Comment by jdukleth** [comment]
Looks great. I think this should be baked into Apigility.

File: src/HttpMethodOverrideListener.php [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+     * @param array $httpMethodOverride
+     */
+    public function __construct(array $httpMethodOverride)
+    {
+        $this->httpMethodOverride = $httpMethodOverride;
+    }
+
+    /**
+     * Priority is set very high (should be executed before all other listeners that rely on the request method value).
+     * TODO: Check priority value, maybe value should be even higher??
```
Comment (weierophinney) [inline src/HttpMethodOverrideListener.php]: This is an appropriate value. The OPTIONS handling (i.e., HTTP method negotiation) happens at -100. Currently, we already have versioning (for Accept headers) happening at -40 (route-based version checking happens at -41). The main thing is that it happens before the OPTIONS handling, so this works fine.

File: src/Module.php [hunk 18]
```diff
@@ -48,6 +48,11 @@ public function onBootstrap(MvcEvent $e)
         $services->get(AcceptFilterListener::class)->attach($eventManager);
         $services->get(ContentTypeFilterListener::class)->attach($eventManager);
 
+        $contentNegotiationOptions = $services->get(ContentNegotiationOptions::class);
+        if ($contentNegotiationOptions->getXHttpMethodOverrideEnabled()) {
+            $services->get(HttpMethodOverrideListener::class)->attach($eventManager);
```
Comment (weierophinney) [inline src/Module.php]: I very much like this conditional attachment.

**Comment by weierophinney** [comment]
Thanks, @Wilt\!

**Comment by Wilt** [comment]
You are very welcome :smiley:

**Pull request closed without merging**